clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3675,14 +4062,260 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     removeFieldMemberExpr(FD, ArrayType);
     return true;
   }
+  using SyclKernelFieldHandler::enterArray;
+  using SyclKernelFieldHandler::enterStruct;
+  using SyclKernelFieldHandler::handleNonDecompStruct;
+  using SyclKernelFieldHandler::handlePointerType;
+  using SyclKernelFieldHandler::handleScalarType;
+  using SyclKernelFieldHandler::handleSyclSpecialType;
+  using SyclKernelFieldHandler::handleUnionType;
+  using SyclKernelFieldHandler::leaveArray;
+  using SyclKernelFieldHandler::leaveStruct;
+};
+
+class FreeFunctionKernelBodyCreator : public SyclKernelFieldHandler {
+  SyclKernelDeclCreator &DeclCreator;
+  llvm::SmallVector<Stmt *, 16> BodyStmts;
+  FunctionDecl *FreeFunc;
+  SourceLocation KernelCallerSrcLoc; // KernelCallerFunc source location.
```

Comment:
Does kernel caller make sense in this context? I would just rename this as `FreeFuncSrcLoc`